### ISSUE TITLE
Datastore fails to sync over missing model

### ISSUE THREAD

**Issue by naripok** [aws-amplify/amplify-js#7022]
**Describe the bug**
1 - I have a many-to-many relationship
2 - I trigger a save event on my models using datastore save method
3 - Something happens and a 'child' model fails to be saved after the 'link' model have been saved

Now, datastore fails to sync at all, not only the affected model, but all models, over all users, at any location.

The failed graphql sync query returns a 200 status code with the error:
`Cannot return null for non-nullable type: 'Component' within parent 'ChildComponent' (/syncChildComponents/items[757]/parent)`

The error is not captured by the error handler callback defined on datastore configuration, nor a hub event is dispatched informing me about the error.

Nothing really happens, datastore fails to work, silently, until I try to dispatch a  (complex) save operation just to find out that my db is in an inconsistent state and the work will not be synced.

Clearing the local does not have any effect as datastore tries to sync over the, now inconsistent, db state and gets stuck.

 The only fix that worked for me until now is to clear my entire dynamo table, which is not really viable.

**To Reproduce**

1. Save a parent model to datastore
2. Save a link model pointing to a parent model and a child model that will never be saved
3. Do not save the correspondent child model
4. Try to sync datastore with `Datastore.start()`
5. Cry tears of blood...

**Expected behavior**
Datastore should be robust to inconsistent entries, skip the failed models and sync all other data.
It should, preferentially, warn me about the failed model, either on the error handler callback and / or by dispatching a hub event over datastore channel.
The event should contain enough information so I can take the appropriated action in order to make my db consistent again.
It should NOT fail silently.

**Code Snippet**
*schema*
```
Component @model
{
  id: ID!
  children: [ChildComponent]  @connection(keyName: "byChild", fields: ["id"])
}

ChildComponent @model(queries: null)
@key(name: "byChild", fields: ["parentID", "childID"])
@key(name: "byParent", fields: ["childID", "parentID"])
{
  id: ID!
  parentID: ID!
  childID: ID!
  parent: Component! @connection(fields: ["parentID"])
  child: Component! @connection(fields: ["childID"])
}
```
Please, give me some advice on a better temporary solution.
Removing datastore from my application would be a big pain at this point and I cannot risk losing data every time a client have a internet failure.

Thanks.

EDIT:

As a workaround, for visibility, I'm now calling `Datastore.start()` every time my application starts and maintaining a flag synced with hub events. If no `ready` event is dispatched, I know Datastore sync is not working.
However, this does not solves the DB inconsistency problems.
Is there any way for me to find out which record is missing without performing a full, multiple table, scan?

Cheers.

**Comment by iartemiev** [comment]
@naripok could you please share your app code with how you're saving these models

**Comment by naripok** [comment]
Sure!

*schema*
```
...

Component @model
{
  id: ID!
  children: [ChildComponent]  @connection(keyName: "byChild", fields: ["id"])
}

ChildComponent @model(queries: null)
@key(name: "byChild", fields: ["parentID", "childID"])
@key(name: "byParent", fields: ["childID", "parentID"])
{
  id: ID!
  parentID: ID!
  childID: ID!
  parent: Component! @connection(fields: ["parentID"])
  child: Component! @connection(fields: ["childID"])
}

...
```

*Save method implementation*
```
export const saveAppComponentTree =
  ({app = {}, data = {}, theme = {}, parentData = {}, componentTree = {}}) =>
    F.go(function*() {

      let savedRoot
      let savedTheme
      let savedContent
      let savedStyles
      let savedLink
      let savedChildren

      const rootData = { path: ''
                       , appID: app.id
                       , url: app.url
                       , ...removeEmptyProperties (data)
                       }
      const { name
            , children
            , styles
            , content
            } = componentTree

      savedRoot = yield createComponent ({...rootData, name})

      if (!S.equals ({}) (viewParent (parentData))) {
        savedLink = yield createParentChildLink (savedRoot) (parentData)
      }

      const withParent = S.insert ('parentID')
      const createLinkedData =
        parent => createModel =>
        S.pipe
          ([ S.map (withParent (viewId (parent)))
           , S.map (createModel)
           , F.resolveParallelArray
          ])

      const createRootModelData = createLinkedData (savedRoot)

      savedContent = yield createRootModelData (createContent) (content)
      savedStyles = yield createRootModelData (createStyle) (styles)
      if (theme) {
        savedTheme = yield createRootModelData
          (createTheme)
          ([{json: JSON.stringify (theme)}])
      }

      savedChildren = yield F.resolveParallelArray
        (S.mapWithIndex
          (p => saveAppComponentTree
            ({app, parentData: {parent: savedRoot, key: S.fst (p)}, componentTree: S.snd (p)})
          )
          (children)
        )

      return {
        savedRoot,
        savedChildren,
        savedTheme,
        savedContent,
        savedStyles,
        savedLink,
      }
    })

const createContent =
  data => save (new models.Content (data))

const createStyle =
  data => save (new models.Style (data))

const createTheme =
  data => save (new models.Theme (data))

const createComponent =
  data => save (new models.Component (data))

const createParentChildLink =
  child => parentData =>
    save (new models.ChildComponent
      ({ child
       , ...parentData
       })
    )

```

The interesting bits here are:
```
({..., parentData = {}, componentTree = {}}) =>

...

      savedRoot = yield createComponent ({...rootData, name})

      if (!S.equals ({}) (viewParent (parentData))) {
        savedLink = yield createParentChildLink (savedRoot) (parentData)
      }

...

      savedChildren = yield F.resolveParallelArray
        (S.mapWithIndex
          (p => saveAppComponentTree
            ({app, parentData: {parent: savedRoot, key: S.fst (p)}, componentTree: S.snd (p)})
          )
          (children)
        )

...

const createComponent =
  data => save (new models.Component (data))

const createParentChildLink =
  child => parentData =>
    save (new models.ChildComponent
      ({ child
       , ...parentData
       })
    )
```
This is a tree structure, and I save it to datastore with a recursive function.
`S` is for `sanctuary` and `F` is for `fluture`

If further clarification is needed, I'm happy to help.
Cheers!

**Comment by iartemiev** [comment]
What's an example of `data`, `child`, and `parentData` that's actually getting passed to the `save` method here:
```js
const createComponent =
  data => save (new models.Component (data))

const createParentChildLink =
  child => parentData =>
    save (new models.ChildComponent
      ({ child
       , ...parentData
       })
    )
```

**Comment by naripok** [comment]
Sorry,

*data*
```
{
  "path": "button",
  "gtag": "tag",
  "fbq": "",
  "title": "Who is Father Julius?",
  "description": "An simple action website with a button.",
  "appVersion": "0.2.0"
}
```
*child*
```
{
  "path": "button",
  "appID": "bf5a90f0-1ec9-4e09-957a-e3278d9570cf",
  "url": "domain",
  "gtag": "tag",
  "title": "Who is Father Julius?",
  "description": "An simple action website with a button.",
  "appVersion": "0.2.0",
  "name": "BackgroundImg",
  "id": "64cfc769-1a75-4b74-96fc-692141f17dd9"
}
```

*parentData*
```
{
  "parent": {
    "path": "button",
    "appID": "bf5a90f0-1ec9-4e09-957a-e3278d9570cf",
    "url": "domain",
    "gtag": "tag",
    "title": "Who is Father Julius?",
    "description": "An simple action website with a button.",
    "appVersion": "0.2.0",
    "name": "BackgroundImg",
    "id": "64cfc769-1a75-4b74-96fc-692141f17dd9"
  },
  "key": 0
}
```
Just to clarify, the `data` and `child` are send on the first iteration of the method, and `parentData` is used on the second iteration.
`child` and `parent` on `parentData` are actually DataStore Model objects, but when I copy then from chrome they rasterizes on  plain objects.

Sorry, may I ask a dumb question?
I was implementing a scanning method for fixing the DB state, and I just realized I can't do this because graphql calls does not really delete any entry, just update the _deleted flag, and Datastore doesn't care about the flag and still on stale state.
Is there a way to delete entries programmatically?

**Comment by wei** [comment]
@naripok Is there anyway you could provide a minimal reproducible example repo from scratch using the latest amplify CLI `4.30.0` and latest node packages? It will help us tremendously! ❤️

**Comment by naripok** [comment]
Hi, @wei,

Take this one by [example](https://github.com/naripok/datastore7022).

I tried to reproduce with the offline network sim on dev tools, but didn't have any luck.
The easiest way to reproduce the bug is to run the client, wait for datastore to finish syncing the models, then you delete some of then directly from the dynamo table.

As my application saves 1k-2k models at a time, and a lot of then are not direct parent-child, but data contained on the nodes, and considering the fact that this bug affected me like 6 or 7 times by now, I suspect the problem should be correlated with the order that the models are being synced and the time between the syncing of the link and the relevant nodes.

For instance, when chrome shows any problems and closes itself, as sometimes happens, and the indexedDB state gets messed, this problem arises.

I realize this was not much of a help, so I'm sorry.

**Comment by naripok** [comment]
Hey guys,

I tried my best to reproduce the bug within a minimal implementation, however I had no success.
The only way I found to reproduce it reliably without inducing a browser segfault is to mess directly with dynamo.
As it is not the observed buggy behavior and adds no information to the matter, I'll close this issue.
My application works reliably using graphql calls, and also gets improved performance from the requests parallelism, so I'll just move out of DataStore for now.

Thx for the help!
Cheers.
